### PR TITLE
Added Elastic Table Samples

### PR DESCRIPTION
**Pull request by kalavany**
Added Elastic Table Samples

### CODE REVIEW

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+                // Delete all the data that is created.
+                DeleteMultipleRecords(client, deviceId, createdRecordIds);
+            }
+            finally
+            {
+                Utility.DeleteSensorDataEntity(client);
+            }
+        }
+
+        private static Guid CreateRecord(IOrganizationService organizationService, string deviceId, ref string sessionToken)
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs:54]: Please change `organizationService` to `service` everywhere.

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+
+                EnergyConsumption energyConsumption = new()
+                {
+                    Voltage = i,
+                    VoltageUnit = "Volts",
+                    Power = i + 1,
+                    PowerUnit = "Watts"
+                };
+
+                entity["contoso_energyconsumption"] = JsonConvert.SerializeObject(energyConsumption);
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs:281]: Can you add a comment to call out that `contoso_energyconsumption` is a string column with JSON format?

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+            {
+                RequestName = "DeleteMultiple",
+                ["Targets"] = new EntityReferenceCollection(entityReferences)
+            };
+            organizationService.Execute(deleteMultipleRequest);
+
+            Console.WriteLine($"Deleted {entityReferences.Count} {"contoso_sensordata"} records.\n");
+        }
+
+        private class EnergyConsumption
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs:448]: Can you add a comment that this class is used for the JSON column?

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        private static string SensorTypeSchemaName = "contoso_SensorType";
+        private static string ValueSchemaName = "contoso_Value";
+        private static string TimeStampSchemaName = "contoso_TimeStamp";
+        private static string EnergyConsumptionSchemaName = "contoso_EnergyConsumption";
+        private static string SensorDataLogicalName = SensorDataSchemaName.ToLower();
+
+        internal static void CreateSensorDataEntity(ServiceClient client)
+        {
+            Console.WriteLine($"Creating {SensorDataSchemaName} entity.. \n");
+            EntityMetadata entityMetadata = new EntityMetadata
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs:22]: You don't seem to be using this `entityMetadata` anywhere.
I know we want to use it in the future.
Can you remove or add comment about why it is here?

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+                DisplayCollectionName = new Label("SensorData", 1033),
+                Description = new Label("Stores IoT data emitted from devices", 1033),
+                OwnershipType = OwnershipTypes.UserOwned,
+                // TableType = "Elastic",
+                IsActivity = false,
+                CanCreateCharts = new BooleanManagedProperty(false),
+            };
+
+            AttributeMetadata primaryAttributeMetadata = new AttributeMetadata
+            {
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs:35]: Aame with this AttributeMetadata. You aren't using it yet, are you?

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+                            },
+                            { "MaxLength", 100 }
+                        }
+                    }
+                }
+            };
+
+            Dictionary<string, List<string>> customHeaders = new Dictionary<string, List<string>>();
+            customHeaders["Content-Type"] = new List<string>() { "application/json" };
+            client.ExecuteWebRequest(HttpMethod.Post, $"EntityDefinitions", entityMetadataObject.ToString(), customHeaders);
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs:167]: `$"EntityDefinitions"`
$ not needed

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+                    Description = new Label("Stores unstructured energy consumption data as reported by device", 1033)
+                }
+            };
+
+            client.Execute(createSensorDataAttributeRequest);
+            Console.WriteLine($"Created {EnergyConsumptionSchemaName} attribute.\n");
+        }
+
+        internal static void DeleteSensorDataEntity(ServiceClient client)
+        {
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs:231]: RE: 
> `DeleteSensorDataEntity(ServiceClient client)`

s/b `DeleteSensorDataEntity(IOrganizationService service)`

For the `CreateSensorDataEntity` method I understand using `ServiceClient` b/c you need to use `ExecuteWebRequest`. 

But we should use `IOrganizaitonService` any other time.

**Review by JimDaly — CHANGES_REQUESTED** [review]
Looks great. Found a few things to change.

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs (hunk 4 above)
Comment (kalavany) [reply dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Utility.cs:22]: yes, my bad, forgot to remove it.

File: dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+
+            Console.WriteLine($"{"contoso_sensordata"} record created.\n");
+
+            // Capture the session token
+            sessionToken = response.Results["x-ms-session-token"].ToString();
+
+            return response.id;
+        }
+
+        private static void UpdateRecord(IOrganizationService service, Guid contosoSensorDataId, string deviceId, ref string sessionToken)
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/ElasticTables/ElasticTableSamples/ElasticTableSamples/Program.cs:86]: @kalavany 
Sent you a question about this.

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/Settings.cs [hunk 9]
```diff
@@ -1,18 +1,26 @@
...
 {
-    public static class Settings {
+    public static class Settings
+    {
 
         /// <summary>
         /// The number of records to create for all samples in this solution.
         /// </summary>
-        public const int NumberOfRecords = 100;
+        public const int NumberOfRecords = 100000;
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/xMultipleSamples/Settings.cs:9]: Please leave this at 100. People can increase it if they want more

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/Settings.cs [hunk 10]
```diff
@@ -1,18 +1,26 @@
...
         /// </summary>
-        public const short BatchSize = 1000;
+        public const short StandardBatchSize = 1000;
+
+        /// <summary>
+        /// The maximum number of records operations to send with 
+        /// CreateMultiple, UpdateMultiple and DeleteMultiple
+        /// for Elastic entities cannot exceed 100.
+        /// </summary>
+        public const short ElasticBatchSize = 100;
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/xMultipleSamples/Settings.cs:23]: I don't remember seeing this mentioned in the docs.

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/Utility.cs [hunk 11]
```diff
@@ -276,34 +275,44 @@ public static void CreateExampleTable(IOrganizationService service, string table
...
-            };
+                    Entity = new EntityMetadata
+                    {
+                        SchemaName = tableSchemaName,
+                        DisplayName = new Label("Example", 1033),
+                        DisplayCollectionName = new Label("Examples", 1033),
+                        Description = new Label("A table for code samples.", 1033),
+                        OwnershipType = OwnershipTypes.UserOwned,
+                        IsActivity = false,
+                        DataProviderId = tableType == "Elastic" ? new Guid("1d9bde74-9ebd-4da9-8ff5-aa74945b9f74") : null,
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/xMultipleSamples/Utility.cs:297]: Do we want customers to set this specific GUID for DataProviderId?
I didn't see this mentioned in the docs..

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/appsettings.json [hunk 12]
```diff
@@ -2,9 +2,10 @@
   "ConnectionStrings": {
     "default": "AuthType=OAuth;Url=https://myorg.crm.dynamics.com;Username=someone@myorg.onmicrosoft.com;RedirectUri=http://localhost;AppId=51f81489-12ee-4a9e-aaae-a2591f45987d;LoginPrompt=Auto"
   },
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/xMultipleSamples/appsettings.json:4]: I understand that it probably makes sense to have settings in appssettings.json, but all the other settings to control the behavior of the sample are in the Settings.cs class.

Can you move this setting there?

Also, I want the default value to be false. 
If someone wants to test elastic tables, they can set it to true.

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/ElasticUtility.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+                { "CanCreateCharts", new JObject()
+                    {
+                        { "Value", false },
+                        { "CanBeChanged", true },
+                        { "ManagedPropertyLogicalName", "cancreatecharts" }
+                    }
+                },
+                { "HasActivities", false },
+                { "HasNotes", false },
+                { "Attributes", new JArray()
```
Comment (JimDaly) [inline dataverse/orgsvc/C#-NETCore/xMultipleSamples/ElasticUtility.cs:75]: I don't see the `sample_Description` column added here like the others.

This sample supports another sample about writing plug-ins, so the tables and data need to be the same.

**Review by JimDaly — CHANGES_REQUESTED** [review]
This sample supports another sample about writing plug-ins for CreateMultiple and UpdateMultiple.  It should work regardless of the type of table.

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/Utility.cs (hunk 11 above)
Comment (kalavany) [reply dataverse/orgsvc/C#-NETCore/xMultipleSamples/Utility.cs:297]: missed reverting this

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/ElasticUtility.cs (hunk 13 above)
Comment (kalavany) [reply dataverse/orgsvc/C#-NETCore/xMultipleSamples/ElasticUtility.cs:75]: this is only for creating the table, the other common attributes are getting created in utility itself

File: dataverse/orgsvc/C#-NETCore/xMultipleSamples/Settings.cs (hunk 10 above)
Comment (kalavany) [reply dataverse/orgsvc/C#-NETCore/xMultipleSamples/Settings.cs:22]: This is the recommended batch size for xMultiple, we might have to document this.

**Comment by JimDaly** [comment]
@kalavany 
I've added the functionality this provides into the samples.

**Pull request closed without merging**